katpoint/projection.py

```diff
@@ -129,24 +133,196 @@
+
+    @classmethod
+    def __enter__(cls):
+        pass
+
+    @classmethod
+    def __exit__(cls, *args, **kws):
+        cls._treatment = cls._old_treatment
+
+    @classmethod
+    def get_treatment(cls):
+        """The current treatment of out-of-range values."""
+        return cls._treatment
+
+    @classmethod
+    def set_treatment(cls, treatment):
+        """"Change the treatment of out-of-range values.
+
+        The supported treatments are:
+          - 'raise': raise :class:`OutOfRangeError` (the default)
+          - 'nan': replace out-of-range values with NaNs
+          - 'clip': replace out-of-range values with nearest valid values
+
+        Parameters
+        ----------
+        treatment : {'raise', 'nan', 'clip'}
+            New treatment
+
+        Returns
+        -------
+        context_manager : :class:`OutOfRange`
+            An instance intended to be used in a with-statement
+
+        Raises
+        ------
+        ValueError
+            If `treatment` is not a recognised option
+        """
+        valid_treatments = {'raise', 'nan', 'clip'}
+        if treatment not in valid_treatments:
+            raise ValueError("Unknown out-of-range treatment '{}', must be one of {}"
+                             .format(treatment, valid_treatments))
+        cls._old_treatment = cls._treatment
+        cls._treatment = treatment
+        return cls()
+
+    @classmethod
+    def treat(cls, x, err_msg, lower=None, upper=None):
+        """Apply treatment to any out-of-range values in `x`.
+
+        Parameters
+        ----------
+        x : real number or array-like of real numbers
+            Input values (left untouched)
+        err_msg : string
+            Error message passed to exception if treatment is 'raise'
+        lower, upper : real number or None, optional
+            Bounds for values in `x` (specify at least one bound!)
+
+        Returns
+        -------
+        treated_x : float or array of float
+            Treated values (guaranteed to be in range or NaN)
+
+        Raises
+        ------
+        OutOfRangeError
+            If any values in `x` are out of range and treatment is 'raise'
+
+        Notes
+        -----
+        If a value is out of bounds by less than the machine precision, it is
+        considered a rounding error. It is not treated as out-of-range to avoid
+        false alarms, but instead silently clipped to ensure that all returned
+        data is in the valid range.
+        """
+        clipped_x = np.asarray(np.clip(x, lower, upper), dtype=np.float)
+        # Suppress false alarms due to rounding errors -> only flag substantial outliers
+        out_of_range = np.abs(x - clipped_x) > np.finfo(float).eps
+        if cls._treatment == 'raise' and np.any(out_of_range):
+            raise OutOfRangeError(err_msg)
+        elif cls._treatment == 'nan':
+            clipped_x[out_of_range] = np.nan
+        return clipped_x if clipped_x.ndim else clipped_x.item()
```

Comment:
Yes, that is exactly the idea.

To me, 0d arrays are simply ephemeral constructs that want to be scalars. This is how easy it is to lose one:
```
>>> import numpy as np                                                           
>>> x = np.array(2)
>>> x                                              
array(2)
>>> np.cos(x)                                                     
-0.4161468365471424
>>> x ** 2
4
```
So if NumPy thinks nothing of dropping it, why should I? 😆 

Plus, it's the only nice way to support scalars + arrays with the same code that I could find. I would be very interested in an improvement 😉

Reply:
I've reconsidered this and found an improvement :-)

Instead of checking `ndim` of the output variable, I now run `np.isscalar` on the input variable. This way I can preserve scalars and 0d arrays.

Reply:
Here is some required reading on the subject: https://stackoverflow.com/a/794812 😄